code/modules/vehicles/mecha/equipment/tools/work_tools.dm

```diff
@@ -158,67 +158,72 @@
+		turflist += get_step(get_step(chassis, carddir), turn(carddir, 90))
 
-	if(reagents.total_volume <= 0)
-		return
-	playsound(chassis, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
+	for(var/turf/targetturf in turflist)
+		var/obj/effect/particle_effect/water/extinguisher/water = new /obj/effect/particle_effect/water/extinguisher(targetturf)
+		var/datum/reagents/water_reagents = new /datum/reagents(required_amount/8) //required_amount/8, because the water usage is split between eight sprays. As of this comment, required_amount/8 = 10u each.
+		water.reagents = water_reagents
+		water_reagents.my_atom = water
+		reagents.trans_to(water, required_amount/8)
+		water.move_at(get_step(chassis, get_dir(targetturf, chassis)), 2, 4) //Target is the tile opposite of the mech as the starting turf.
 
-	sprays_left += 5
-	add_hiddenprint(source) //log prints so admins can figure out who touched it last.
-	log_combat(source, target, "fired an extinguisher at")
-	spray_extinguisher(target)
-	return ..()
+	playsound(chassis, 'sound/effects/extinguish.ogg', 75, TRUE, -3)
 
-/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/spray_extinguisher(atom/target)
-	var/direction = get_dir(chassis, target)
-	var/turf/T1 = get_turf(target)
-	var/turf/T2 = get_step(T1,turn(direction, 90))
-	var/turf/T3 = get_step(T1,turn(direction, -90))
-	var/list/targets = list(T1,T2,T3)
-
-	var/obj/effect/particle_effect/water/extinguisher/water = new /obj/effect/particle_effect/water/extinguisher(get_turf(chassis))
-	var/datum/reagents/water_reagents = new /datum/reagents(5)
-	water.reagents = water_reagents
-	water_reagents.my_atom = water
-	reagents.trans_to(water, 1)
-
-	var/delay = 2
-	var/datum/move_loop/our_loop = water.move_at(pick(targets), delay, 4)
-	RegisterSignal(our_loop, COMSIG_PARENT_QDELETING, .proc/water_finished_moving)
-
-/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/water_finished_moving(datum/move_loop/has_target/source)
-	SIGNAL_HANDLER
-	sprays_left--
-	if(!sprays_left)
+/obj/item/mecha_parts/mecha_equipment/extinguisher/proc/attempt_refill(mob/usr)
+	if(reagents.maximum_volume == reagents.total_volume)
+		return
+	var/turf/in_front = get_step(chassis, chassis.dir)
+	var/obj/structure/reagent_dispensers/watertank/refill_source = locate(/obj/structure/reagent_dispensers/watertank) in in_front
+	if(!refill_source)
+		to_chat(usr, span_notice("Refill failed. No compatible tank found."))
+		return
+	if(!refill_source.reagents?.total_volume)
+		to_chat(usr, span_notice("Refill failed. Source tank empty."))
 		return
-	extinguish(source.target)
+
+	refill_source.reagents.trans_to(src, reagents.maximum_volume)
+	playsound(chassis, 'sound/effects/refill.ogg', 50, TRUE, -6)
+	return
 
 /obj/item/mecha_parts/mecha_equipment/extinguisher/get_snowflake_data()
 	return list(
 		"snowflake_id" = MECHA_SNOWFLAKE_ID_EXTINGUISHER,
 		"reagents" = reagents.total_volume,
 		"total_reagents" = reagents.maximum_volume,
+		"minimum_requ" = required_amount,
 	)
```

Comment:
extremely nitpick but usually its req not requ for abbreviating required, no?

Reply:
It was originally "minimum_required" but I just started deleting characters to get under the 80 char limit per line that js files have.

Reply:
That is so mood I hate that